Sources/Paywalls/Components/PaywallComponentBases.swift

```diff
@@ -0,0 +1,374 @@
+public typealias LocaleId = String
+public typealias ColorHex = String
+
+public typealias DisplayString = PaywallComponent.LocaleResources<String>
+
+extension ColorHex {
+
+    enum Error: Swift.Error {
+
+        case invalidStringFormat(String)
+        case invalidColor(String)
+
+    }
+
+    public func toColor() throws -> Color {
+        let red, green, blue, alpha: CGFloat
+
+        guard self.hasPrefix("#") else {
+            throw Error.invalidStringFormat(self)
+        }
+
+        let start = self.index(self.startIndex, offsetBy: 1)
+        let hexColor = String(self[start...])
+
+        guard hexColor.count == 6 || hexColor.count == 8 else {
+            throw Error.invalidStringFormat(self)
+        }
+
+        let scanner = Scanner(string: hexColor)
+        var hexNumber: UInt64 = 0
+
+        if scanner.scanHexInt64(&hexNumber) {
+            // If Alpha channel is missing, it's a fully opaque color.
+            if hexNumber <= 0xffffff {
+                hexNumber <<= 8
+                hexNumber |= 0xff
+            }
+
+            red = CGFloat((hexNumber & 0xff000000) >> 24) / 255
+            green = CGFloat((hexNumber & 0x00ff0000) >> 16) / 255
+            blue = CGFloat((hexNumber & 0x0000ff00) >> 8) / 255
+            alpha = CGFloat(hexNumber & 0x000000ff) / 255
+
+            return .init(red: red, green: green, blue: blue, opacity: alpha)
+        } else {
+            throw Error.invalidColor(self)
+        }
+    }
+
+}
+
+protocol PaywallComponentBase: Codable, Sendable, Hashable, Equatable {
+
+}
+
+public enum PaywallComponent: PaywallComponentBase {
+
+    public func encode(to encoder: any Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+
+        switch self {
+        case .text(let component):
+            try container.encode(ComponentType.text, forKey: .type)
+            try component.encode(to: encoder)
+        case .image(let component):
+            try container.encode(ComponentType.image, forKey: .type)
+            try component.encode(to: encoder)
+        case .spacer(let component):
+            try container.encode(ComponentType.spacer, forKey: .type)
+            try component.encode(to: encoder)
+        case .stack(let component):
+            try container.encode(ComponentType.stack, forKey: .type)
+            try component.encode(to: encoder)
+        case .linkButton(let component):
+            try container.encode(ComponentType.linkButton, forKey: .type)
+            try component.encode(to: encoder)
+        }
+
+    }
+
+    case text(TextComponent)
+    case image(ImageComponent)
+    case spacer(SpacerComponent)
+    case stack(StackComponent)
+    case linkButton(LinkButtonComponent)
+
+    enum CodingKeys: String, CodingKey {
+
+        case type
+
+    }
+
+    enum ComponentType: String, Codable {
+
+        case text
+        case image
+        case spacer
+        case stack
+        case linkButton = "link_button"
```

Comment:
It is using the snake decoding strategy - this is actually here so that it can be initialized using "link_button" for raw representable initializer. I think we can get rid of this by changing the LinkButtonComponent's type to "linkButton", and will change if so.